hubble/src/indexer/api.rs

```diff
@@ -0,0 +1,149 @@
+use std::{fmt::Display, ops::Range};
+
+use axum::async_trait;
+use color_eyre::eyre::Report;
+use futures::Stream;
+use sqlx::Postgres;
+use time::OffsetDateTime;
+use tokio::task::JoinSet;
+use tracing::error;
+
+#[derive(Debug, thiserror::Error)]
+pub enum IndexerError {
+    #[error("received unexpected height {0}: expecting {1}")]
+    UnexpectedHeightSingle(BlockHeight, BlockHeight),
+    #[error("received unexpected height {0} (range {1}): expecting {2}")]
+    UnexpectedHeightRange(BlockHeight, BlockInterval, BlockHeight),
+    #[error("error reading block {0} (range {1}): {2}")]
+    ErrorReadingBlock(BlockHeight, BlockInterval, Report),
+    #[error("expected to receive block {0} (range {1})")]
+    MissingBlock(BlockHeight, BlockInterval),
+    #[error("received block while not expecting more (range {0}): {1}")]
+    TooManyBlocks(BlockInterval, BlockReference),
+    #[error("received error while not expecting more (range {0}): {1}")]
+    TooManyBlocksError(BlockInterval, Report),
+    #[error("no block at: {0}")]
+    NoBlock(BlockSelection),
+    #[error("database error: {0}")]
+    DatabaseError(sqlx::Error),
+    #[error("provider error: {0}")]
+    ProviderError(Report),
+    #[error("internal error: {0}")]
+    InternalError(Report),
+}
+
+impl From<Report> for IndexerError {
+    fn from(error: Report) -> Self {
+        Self::InternalError(error)
+    }
+}
+
+impl From<sqlx::Error> for IndexerError {
+    fn from(error: sqlx::Error) -> Self {
+        Self::DatabaseError(error)
+    }
+}
+
+pub type IndexerId = String;
+pub type BlockHeight = u64;
+pub type BlockHash = String;
+pub type BlockTimestamp = OffsetDateTime;
+pub type BlockRange = Range<BlockHeight>;
+
+#[derive(Clone, Debug)]
+pub struct BlockInterval(BlockRange);
+
+pub trait BlockRangeUtil {
+    fn to_interval(&self) -> BlockInterval;
+    fn len(&self) -> u64;
+}
+
+impl BlockRangeUtil for BlockRange {
+    fn to_interval(&self) -> BlockInterval {
+        BlockInterval(self.clone())
+    }
+
+    fn len(&self) -> u64 {
+        self.end - self.start
+    }
+}
```

Comment:
any reason this is a trait instead of implemeted directly on the BlockInterval type?

Reply:
no, leftover before introducing BlockInterval. I'll fix